File: C++/Lab2/Lab2_task1.cpp
#include <iostream>
using namespace std;

struct Node {                                           // Структура узла
    string val; 
    Node* next;

    Node(string _val) : val(_val), next(nullptr) {}     // Конструктор для вызова
};

struct list {                                           // Структура односвязного списка
    Node* first;
    Node* last;

    list() : first(nullptr), last(nullptr) {}           // Базовый конструктор

    bool is_empty() {                                   // Функция, проверяющая пустоту списка
        return first == nullptr;
    }

    void push_back(string _val) {                       // Функция, отвечающая за добавление элемента в конец списка
        Node* p = new Node(_val);
        if (is_empty()) {
            first = p;
            last = p;
            return;
        }
        last->next = p;
        last = p;
    }

    void print() {                                      // Функция для печати элементов списка
        if (is_empty()) return;
        Node* p = first;
        while (p) {
            cout << p->val << " ";
            p = p->next;
        }
        cout << endl;
    }
};

int main() {                            
    list l;
    cout << l.is_empty() << endl;
    l.push_back("1");
    l.push_back("2");
    l.push_back("3");
    l.push_back("4");
    l.push_back("5");
    cout << l.is_empty() << endl;
    l.print();
    return 0;
}

File: Java/Lab2/Lab2_task1.java
// Структура узла
class Node {
    String val;
    Node next;

    // Конструктор для вызова
    Node(String _val) {
        val = _val;
        next = null;
    }
}

// Структура односвязного списка
class List {
    Node first;
    Node last;

    // Базовый конструктор
    List() {
        first = null;
        last = null;
    }

    // Функция, проверяющая пустоту списка
    boolean is_empty() {
        return first == null;
    }

    // Функция, отвечающая за добавление элемента в конец списка
    void push_back(String _val) {
        Node p = new Node(_val);
        if (is_empty()) {
            first = p;
            last = p;
            return;
        }
        last.next = p;
        last = p;
    }

    // Функция для печати элементов списка
    void print() {
        if (is_empty()) return;
        Node p = first;
        while (p != null) {
            System.out.print(p.val + " ");
            p = p.next;
        }
        System.out.println();
    }
}

public class Main {
    public static void main(String[] args) {
        List l = new List();
        System.out.println(l.is_empty());
        l.push_back("1");
        l.push_back("2");
        l.push_back("3");
        l.push_back("4");
        l.push_back("5");
        System.out.println(l.is_empty());
        l.print();
    }
}
